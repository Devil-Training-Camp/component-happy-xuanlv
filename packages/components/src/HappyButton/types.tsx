import type React from 'react';
import type { SizeType } from '../types';

export interface HappyButtonProps<T extends keyof JSX.IntrinsicElements = 'button'>
  extends Omit<React.AllHTMLAttributes<HTMLElement>, 'size'> {
  /**
   * 将按钮宽度调整为其父宽度的选项
   */
  block?: boolean;
  /**
   * 禁用状态
   */
  disabled?: boolean;
  /**
   * 是否有边框
   * @defaultValue true
   */
  bordered?: boolean;
  /**
   * 是否为激活状态
   */
  active?: boolean;
  /**
   * 设置按钮的图标组件
   */
  icon?: React.ReactNode;
  /**
   * loading
   */
  loading?: boolean;
  /**
   * 按钮形状
   */
  shape?: 'default' | 'circle' | 'round';
  /**
   * 按钮大小
   * @defaultValue 'middle'
   */
  size?: SizeType;
  /**
   * 按钮类型
   * @defaultValue 'default'
   */
  type?: 'default' | 'primary' | 'dashed';
  /**
   * 是否描边
   */
  outline?: boolean;
  /**
   * 设置危险按钮
   */
  danger?: boolean;
  /**
   * 按钮标签名
   * @defaultValue 'button'
   */
  buttonTagName?: T | string;
}
