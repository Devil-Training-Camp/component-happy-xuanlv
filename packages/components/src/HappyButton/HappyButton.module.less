@import '../style/variable-default';

.happyButton {
  position: relative;
  display: inline-flex;
  flex: none;
  align-items: center;
  justify-content: center;
  height: @happy-btn-size-base;
  padding: 0 @happy-btn-padding-horizontal-base;
  font-size: @happy-font-size-base;
  font-weight: @happy-btn-font-weight;
  line-height: 1;
  color: @happy-btn-default-color;
  white-space: nowrap;
  vertical-align: middle;
  touch-action: manipulation;
  cursor: pointer;
  user-select: none;
  background: @happy-btn-default-bg;
  border: @happy-btn-border-width @happy-btn-border-style @happy-btn-default-border;
  border-radius: @happy-btn-border-radius-base;
  outline: 0;
  box-shadow: @happy-btn-shadow;
  transition: all 0.3s ease-out;
  appearance: none;

  &:not(.disabled) {
    &:hover,
    &:focus {
      color: @happy-primary-color;
      border-color: @happy-primary-color;
    }

    &:active,
    &:global(.happy-button-active) {
      color: @happy-primary-7;
      border-color: @happy-primary-7;
    }
  }
}

.block {
  width: 100%;
}

.primary:not(.disabled) {
  color: @happy-btn-primary-color;
  background: @happy-primary-color;
  border-color: transparent;
  box-shadow: @happy-btn-primary-shadow;

  &:hover,
  &:focus {
    color: @happy-btn-primary-color;
    background: @happy-primary-5;
    border-color: transparent;
  }

  &:active,
  &:global(.happy-button-active) {
    color: @happy-btn-primary-color;
    background: @happy-primary-7;
  }
}

.dashed {
  border-style: dashed;
}

.outline:not(.disabled) {
  color: @happy-primary-color;
  background: @happy-btn-default-bg;
  border-color: @happy-primary-color;

  &:not(.primary) {
    &:hover {
      color: @happy-primary-5;
      border-color: @happy-primary-5;
    }

    &:active,
    &:global(.happy-button-active) {
      color: @happy-primary-7;
      border-color: @happy-primary-7;
    }
  }
}

.danger:not(.disabled) {
  color: @happy-btn-danger-bg;
  background: transparent;
  border-color: @happy-btn-danger-bg;

  &:hover,
  &:focus {
    color: @happy-error-color-hover;
    border-color: @happy-error-color-hover;
  }

  &:active,
  &:global(.happy-button-active) {
    color: @happy-error-color-active;
    border-color: @happy-error-color-active;
  }

  &.primary {
    color: @happy-btn-danger-color;
    background: @happy-btn-danger-bg;
    border-color: transparent;

    &:hover,
    &:focus {
      background: @happy-error-color-hover;
    }

    &:active,
    &:global(.happy-button-active) {
      background: @happy-error-color-active;
    }
  }
}

.large {
  height: @happy-btn-size-lg;
  padding: 0 @happy-btn-padding-horizontal-lg;
  font-size: @happy-btn-font-size-lg;
}

.small {
  height: @happy-btn-size-sm;
  padding: 0 @happy-btn-padding-horizontal-sm;
  font-size: @happy-btn-font-size-sm;
  border-radius: @happy-btn-border-radius-sm;
}

.circle,
.onlyIcon {
  width: @happy-btn-size-base;
  height: @happy-btn-size-base;
  padding: 0;

  &.large {
    width: @happy-btn-size-lg;
    height: @happy-btn-size-lg;
  }

  &.small {
    width: @happy-btn-size-sm;
    height: @happy-btn-size-sm;
  }
}

.circle {
  padding: 0;
  border-radius: 50%;
}

.onlyIcon {
  .icon {
    margin-right: 0;
  }
}

.round {
  border-radius: @happy-btn-size-lg;
}

.loading {
  cursor: default;

  &::before {
    position: absolute;
    inset: -1px;
    z-index: 1;
    pointer-events: none;
    content: '';
    background: white;
    border-radius: inherit;
    opacity: 0.35;
    transition: opacity 0.2s;
  }
}

.disabled {
  color: @happy-btn-disable-color;
  cursor: not-allowed;
  background: @happy-btn-disable-bg;
  border-color: @happy-btn-disable-border;

  > * {
    pointer-events: none;
  }
}

.borderless {
  border-width: 0;
  box-shadow: none;
}

.icon {
  margin-right: 8px;
}
