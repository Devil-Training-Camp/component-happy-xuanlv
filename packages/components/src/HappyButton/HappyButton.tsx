import React from 'react';
import clsx from 'clsx';
import { HappySpin } from '../HappySpin';
import type { HappyButtonProps } from './types';
import styles from './HappyButton.module.less';

/**
 * HappyButton
 */
function HappyButton<E extends keyof JSX.IntrinsicElements>(
  {
    block,
    className,
    disabled,
    bordered,
    active,
    icon,
    loading,
    shape,
    size = 'middle',
    type,
    outline,
    danger,
    onClick,
    buttonTagName = 'button',
    children,
    ...restProps
  }: HappyButtonProps<E>,
  ref: React.ForwardedRef<HTMLElement>,
) {
  // 标签节点
  const ButtonElement = buttonTagName as unknown as React.HTMLFactory<HTMLElement>;

  function renderIcon() {
    if (loading) {
      return <HappySpin className={styles.icon} style={{ color: 'currentColor' }} size="small" />;
    }
    if (!children) {
      return icon;
    }
    if (icon) {
      return <span className={styles.icon}>{icon}</span>;
    }
    return undefined;
  }

  return (
    <ButtonElement
      className={clsx(styles.happyButton, className, {
        [styles.block]: block,
        [styles.outline]: outline,
        [styles.circle]: shape === 'circle',
        [styles.round]: shape === 'round',
        [styles.primary]: type === 'primary',
        [styles.dashed]: type === 'dashed',
        [styles.large]: size === 'large',
        [styles.small]: size === 'small',
        [styles.onlyIcon]: !children,
        [styles.loading]: loading,
        [styles.disabled]: disabled,
        [styles.borderless]: bordered === false,
        'happy-button-active': active,
        [styles.danger]: danger,
      })}
      ref={ref}
      disabled={disabled}
      type="button"
      onClick={loading ? undefined : onClick}
      {...restProps}
    >
      {renderIcon()}
      {children}
    </ButtonElement>
  );
}

export default React.forwardRef(HappyButton);
