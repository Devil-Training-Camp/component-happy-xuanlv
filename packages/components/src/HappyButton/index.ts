export { default as HappyButton } from './HappyButton';
export * from './types';
