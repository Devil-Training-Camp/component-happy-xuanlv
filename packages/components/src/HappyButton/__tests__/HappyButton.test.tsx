import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import { render, fireEvent } from '@testing-library/react';
import HappyButton from '../HappyButton';
import mountTest from '../../../tests/shared/mountTest';
import snapshotTest from '../../../tests/shared/snapshotTest';

describe('HappyButton', () => {
  mountTest(<HappyButton>按钮</HappyButton>);

  it('正常渲染', () => {
    snapshotTest(<HappyButton>按钮</HappyButton>);
    snapshotTest(<HappyButton size="large">按钮</HappyButton>);
    snapshotTest(<HappyButton size="small">按钮</HappyButton>);
    snapshotTest(<HappyButton outline>按钮</HappyButton>);
    snapshotTest(<HappyButton danger>按钮</HappyButton>);
    snapshotTest(<HappyButton shape="circle">按钮</HappyButton>);
    snapshotTest(<HappyButton shape="round">按钮</HappyButton>);
    snapshotTest(<HappyButton type="primary">按钮</HappyButton>);
    snapshotTest(<HappyButton type="dashed">按钮</HappyButton>);
    snapshotTest(<HappyButton icon={<div>图标</div>}>按钮</HappyButton>);
    snapshotTest(<HappyButton bordered={false}>按钮</HappyButton>);
  });

  it('空按钮', () => {
    const wrapper = render(
      <HappyButton>
        {null}
        {undefined}
      </HappyButton>,
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });

  it('加载时不应单击该按钮', () => {
    const onClick = vi.fn();
    const { container } = render(
      <HappyButton loading onClick={onClick}>
        button
      </HappyButton>,
    );
    fireEvent.click(container.firstChild!);
    expect(onClick).not.toHaveBeenCalledWith();
  });

  it('禁用时不应单击该按钮', () => {
    const onClick = vi.fn();
    const { container } = render(
      <HappyButton disabled onClick={onClick}>
        button
      </HappyButton>,
    );
    fireEvent.click(container.firstChild!);
    expect(onClick).not.toHaveBeenCalledWith();
  });

  it('自定义链接标签', () => {
    const wrapper = render(
      <HappyButton target="_blank" href="https://react.dev/reference/react" buttonTagName="a">
        链接
      </HappyButton>,
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });

  it('loading 状态改变', async () => {
    const { container, rerender, unmount } = render(<HappyButton>按钮</HappyButton>);
    rerender(<HappyButton loading />);
    expect(container.querySelectorAll('.happy-spin').length).toBe(1);
    rerender(<HappyButton loading={false} />);
    expect(container.querySelectorAll('.happy-spin').length).toBe(0);
    expect(unmount).not.toThrow();
  });
});
