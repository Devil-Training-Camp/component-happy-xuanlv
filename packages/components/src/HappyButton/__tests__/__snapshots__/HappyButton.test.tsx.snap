// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HappyButton > 正常渲染 1`] = `
<button
  class="_happyButton_5fc01e"
  type="button"
>
  按钮
</button>
`;

exports[`HappyButton > 正常渲染 2`] = `
<button
  class="_happyButton_5fc01e _large_5fc01e"
  type="button"
>
  按钮
</button>
`;

exports[`HappyButton > 正常渲染 3`] = `
<button
  class="_happyButton_5fc01e _small_5fc01e"
  type="button"
>
  按钮
</button>
`;

exports[`HappyButton > 正常渲染 4`] = `
<button
  class="_happyButton_5fc01e _outline_5fc01e"
  type="button"
>
  按钮
</button>
`;

exports[`HappyButton > 正常渲染 5`] = `
<button
  class="_happyButton_5fc01e _danger_5fc01e"
  type="button"
>
  按钮
</button>
`;

exports[`HappyButton > 正常渲染 6`] = `
<button
  class="_happyButton_5fc01e _circle_5fc01e"
  type="button"
>
  按钮
</button>
`;

exports[`HappyButton > 正常渲染 7`] = `
<button
  class="_happyButton_5fc01e _round_5fc01e"
  type="button"
>
  按钮
</button>
`;

exports[`HappyButton > 正常渲染 8`] = `
<button
  class="_happyButton_5fc01e _primary_5fc01e"
  type="button"
>
  按钮
</button>
`;

exports[`HappyButton > 正常渲染 9`] = `
<button
  class="_happyButton_5fc01e _dashed_5fc01e"
  type="button"
>
  按钮
</button>
`;

exports[`HappyButton > 正常渲染 10`] = `
<button
  class="_happyButton_5fc01e"
  type="button"
>
  <span
    class="_icon_5fc01e"
  >
    <div>
      图标
    </div>
  </span>
  按钮
</button>
`;

exports[`HappyButton > 正常渲染 11`] = `
<button
  class="_happyButton_5fc01e _borderless_5fc01e"
  type="button"
>
  按钮
</button>
`;

exports[`HappyButton > 空按钮 1`] = `
<button
  class="_happyButton_5fc01e"
  type="button"
/>
`;

exports[`HappyButton > 自定义链接标签 1`] = `
<a
  class="_happyButton_5fc01e"
  href="https://react.dev/reference/react"
  target="_blank"
  type="button"
>
  链接
</a>
`;
